bootcompiler/src/main/scala/org/mozartoz/bootcompiler/transform/OnStackMarking.scala

```diff
@@ -0,0 +1,221 @@
+package org.mozartoz.bootcompiler
+package transform
+
+import ast._
+import oz._
+import symtab._
+import scala.collection._
+import com.sun.javafx.fxml.expression.BinaryExpression
+
+sealed class VarState(val onStack: Boolean, val locked: Boolean) {
+}
+object Declared      extends VarState(false, false)
+object Captured      extends VarState(false, false)
+object BoundFirst    extends VarState(true, true)
+object ToInstantiate extends VarState(false, true)
+
+case class VarAndState(v: Variable, state: VarState)
+
+/**
+ * Looks for declarations and the usages of those declarations.
+ * The Transform methods detect local statements and add declared variables to the environment
+ * The Walk methods are collecting their usages and tagging them
+ * 
+ * If a variable is bound after its declaration, we can shortcut it
```

Comment:
If a variable is bound after its declaration without being read, there is no need to create an identity for the variable